### ISSUE TITLE
Manually reload tls certificates

### ISSUE THREAD

**Issue by Nepoxx** [containous/traefik#5495]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

My tls certificates are generated with Let's Encrypt remotely and are used by Traefik through a glusterfs mount. For that reason, Traefik is unable to properly monitor file changes and thus never knows when certificates are renewed (so it will serve an expired certificate). Having a way to tell Traefik to reload new certificates (or file configs in general) would allow the user to circumvent cases when Traefik is unable to use inotify.

#### Traefik Version
2.0

Similar to https://github.com/containous/traefik/issues/1623

**Comment by ldez** [comment]
Hello,

could you give more information about your Traefik configuration?

**Comment by RomRider** [comment]
Same issue for me, I have this configuration in place:
```yaml
tls:
  certificates:
    - certFile: /etc/letsencrypt/live/DOMAIN/fullchain.pem
      keyFile: /etc/letsencrypt/live/DOMAIN/privkey.pem

  stores:
    default:
      defaultCertificate:
        certFile: /etc/letsencrypt/live/DOMAIN/fullchain.pem
        keyFile: /etc/letsencrypt/live/DOMAIN/privkey.pem
```

The path is a NFS volume and the certificates are renewed outside of traefik (because not only used by traefik)

**Comment by ldez** [comment]
could you give more information about your Traefik configuration (static and dynamic configuration)?

**Comment by RomRider** [comment]
Sure:
```yaml
global:
  checkNewVersion: false
  sendAnonymousUsage: false
serversTransport:
  insecureSkipVerify: true

providers:
  docker:
    endpoint: tcp://docker-socket-proxy2:2375
    exposedByDefault: false
    watch: true

  file:
    directory: /configs/
    watch: true

log:
  level: DEBUG
accessLog: {}

api:
  dashboard: true

entryPoints:
  https:
    address: :443
    forwardedHeaders:
      trustedIPs:
        - "127.0.0.1/32"
        - "x.x.x.x"
    proxyProtocol:
      trustedIPs:
        - "127.0.0.1/32"
        - "x.x.x.x"

  publicHttps:
    address: :8443
    forwardedHeaders:
      trustedIPs:
        - "127.0.0.1/32"
        - "x.x.x.x"
    proxyProtocol:
      trustedIPs:
        - "127.0.0.1/32"
        - "x.x.x.x"
```

TLS:
```yaml
tls:
  certificates:
    - certFile: /etc/letsencrypt/live/DOMAIN/fullchain.pem
      keyFile: /etc/letsencrypt/live/DOMAIN/privkey.pem

  stores:
    default:
      defaultCertificate:
        certFile: /etc/letsencrypt/live/DOMAIN/fullchain.pem
        keyFile: /etc/letsencrypt/live/DOMAIN/privkey.pem

  options:
    default:
      minVersion: VersionTLS12
```

Dynamic example:
```yaml
    labels:
      - traefik.enable=true
      - traefik.docker.network=${TRAEFIK_BACKEND_NETWORK}
      - traefik.http.routers.grafana.rule=Host(`grafana.${PRIVATE_DOMAIN}`)
      - traefik.http.routers.grafana.entryPoints=https
      - traefik.http.routers.grafana.tls=true
      - traefik.http.routers.grafana.tls.options=default
      - traefik.http.services.grafana.loadBalancer.server.port=3000
```

**Comment by ldez** [comment]
How you mount the file?

**Comment by RomRider** [comment]
NFS on the host, then I map the folder in the container:
```yaml
    volumes:
      - ./traefik.yml:/traefik.yml:ro
      - ./configs:/configs:ro
      - ${DOCKER_DATA_NFS_FOLDER}/certbot/etc-letsencrypt:/etc/letsencrypt:ro
```

**Comment by Nepoxx** [comment]
For the record, I have pretty much the exact same setup. The only differences I can see is that I used Glusterfs instead of NFS, and I don't mount in `:ro` (read-only).

**Comment by Nepoxx** [comment]
Stack.yml
```stack.yml
version: "3.7"
services:
  traefik:
    image: traefik:v2.0
    ports:
      - "80:80"
      - "443:443"
      - "8080:8080"
      - "8082:8082"
    volumes:
      - type: bind
        source: /var/run/docker.sock
        target: /var/run/docker.sock
      - type: bind
        source: /mnt/gfs/docker/traefik/traefik.yml
        target: /etc/traefik/traefik.yml
      - type: bind
        source: /mnt/gfs/docker/traefik/certs
        target: /certs
      - type: bind
        source: /mnt/gfs/docker/traefik/config
        target: /config
      - type: bind
        source: /mnt/gfs/docker/traefik/acme
        target: /acme
    deploy:
      placement:
        constraints:
          - node.role == manager
    networks:
      traefik-net: {}

networks:
  traefik-net:
    name: traefik-net
    driver: overlay
```

Where `/mnt/gfs` is a gluster mount:
```fstab
localhost:/gfs /mnt/gfs glusterfs defaults,_netdev,backupvolfile-server=venus 0 0
```

traefik.yml
```traefik.yml
entryPoints:
  http:
    address: ":80"

  https:
    address: ":443"

  metrics:
    address: ":8082"

providers:
  docker:
    swarmMode: true
    network: traefik-net
    exposedByDefault: false
  file:
    filename: /config/config.yml
    # inotify does not work with network mounts
    watch: false

# API and dashboard configuration
api:
  insecure: true
  debug: true

metrics:
  prometheus:
    entryPoint: metrics
```

config.yml (mentioned above):
```config.yml
tls:
  certificates:
    - certFile: /certs/_.example.com/_.example.com.cert
      keyFile: /certs/_.example.com/_.example.com.key
http:
  middlewares:
    redirect-to-https:
      redirectScheme:
        scheme: https
  routers:
    http-catchall:
      priority: 0
      entryPoints:
        - http
      middlewares:
        - redirect-to-https@file
      rule: 'hostregexp(`{host:.+}`)'
      service: noop
  services:
    noop:
      loadBalancer:
        servers:
          - url: 'http://127.0.0.1'
```

Replacing `_.example.com.cert` and `_.example.com.key` has no effect on Traefik

**Comment by RomRider** [comment]
An interesting way to handle that would be to provide an API endpoint to reload certs.

**Comment by gtmadev** [comment]
I just ran into this today as well when using a glusterfs volume for file configs. In my main config, I have it watch a directory in the glusterfs volume. That directory is where I store dynamic configurations. Some nodes in docker swarm seem to get the updated file config, and others don't, until I manually restart that instance.

It seems that Traefik is unable to know when files are changed or added (when using the watch option) and the directory is a glusterfs volume.

**Comment by Nepoxx** [comment]
@gtmadev I tried bind mounting using folder and that does not help.

This is a pretty major blocker for me, and I'd like to help move this forward. My initial idea is to add support for a kill signal that would trigger a config reload, probably `SIGHUP` (prometheus does this).

Thoughts?

**Comment by martine-stratdat** [comment]
This is also something I'd like to see.  While Traefik may feel this is covered though it's letsencrypt solution it does not cover all real world cases.  

* This includes easy loading of purchased certificates
* Cases where traefik is just one component of TLS use, including transitioning to traefik from other set ups.
* Instances where rate limiting may be come an issue 

There two solutions that will work here as I see it:
* An API endpoint
* A watch on certificates or certificate entrties

Given that this is a barrier of entry for Traefik I would like to see a solution given a priority.

**Comment by DrEsteban** [comment]
We also desperately need this feature, as the current strategies for triggering certificate reload leave a lot to be desired.

**Comment by hezten** [comment]
I badly need a similar solution too, hope this is getting looked into :-)

**Comment by monotek** [comment]
Same here. 
Using existing secret for bought tls certs in kubernetes and telling traefik via ingress rule what the secretname is.
Updating the secret has no effect at all. 
The old cert is still used...

**Comment by ybizeul** [comment]
Same issue here.

Traefik version 2.2.0 built on 2020-03-25T17:32:57Z

Before changing certificates :
```
drwxr-xr-x    2 root     root          4096 Apr 19 15:56 .
drwxr-xr-x    3 root     root          4096 Apr 19 13:24 ..
-rw-r--r--    1 root     root          1062 Apr 19 15:57 nabox.crt
-rw-r--r--    1 root     root          1679 Apr 19 15:57 nabox.key
-rw-r--r--    1 root     root           127 Apr 19 15:53 traefik.yaml
```

When changing certificates :
```
time="2020-04-19T14:08:16Z" level=debug msg="Configuration received from provider file: {\"http\":{},\"tcp\":{},\"udp\":{},\"tls\":{\"stores\":{\"default\":{}}}}" providerName=file
time="2020-04-19T14:08:16Z" level=info msg="Skipping same configuration" providerName=file
time="2020-04-19T14:08:16Z" level=debug msg="Configuration received from provider file: {\"http\":{},\"tcp\":{},\"udp\":{},\"tls\":{\"stores\":{\"default\":{}}}}" providerName=file
time="2020-04-19T14:08:16Z" level=debug msg="Configuration received from provider file: {\"http\":{},\"tcp\":{},\"udp\":{},\"tls\":{\"stores\":{\"default\":{}}}}" providerName=file
time="2020-04-19T14:08:16Z" level=info msg="Skipping same configuration" providerName=file
time="2020-04-19T14:08:16Z" level=info msg="Skipping same configuration" providerName=file
```

And here is the directory :
```
drwxr-xr-x    2 root     root          4096 Apr 19 15:56 .
drwxr-xr-x    3 root     root          4096 Apr 19 13:24 ..
-rw-r--r--    1 root     root          1062 Apr 19 16:08 nabox.crt
-rw-r--r--    1 root     root          1675 Apr 19 16:08 nabox.key
-rw-r--r--    1 root     root           127 Apr 19 15:53 traefik.yaml
```

Traefik config :
```
      - "--log.level=DEBUG"
      - "--api.insecure=true"
      - "--providers.docker=true"
      - "--providers.docker.exposedbydefault=false"
      - "--providers.file.directory=/tmp/ssl/" << File provider here
      - "--providers.file.watch=true"
      - "--entrypoints.web.address=:80"
      - "--entrypoints.web.http.redirections.entrypoint.to=webssl"
      - "--entrypoints.web.http.redirections.entrypoint.scheme=https"
      - "--entrypoints.webssl.address=:443"
      - "--entrypoints.graphite.address=:2003"
    volumes:               
      - "/var/run/docker.sock:/var/run/docker.sock:ro"            
      - "/opt/conf/ssl:/tmp/ssl" << Traefik config and SSL files
```

**Comment by MartinMuzatko** [comment]
We are generating certificates on request and move them to the right file position. This replaces the inode, which (I believe) breaks traefiks automatic discovery of certificates.

**Comment by majorgearhead** [comment]
Any word on this possible bug?  We programmatically generate our certificates from a known authority and they are replaced for which Traefik cannot see happen.  We have to manually bounce trafik to get it to pick up the new certificates.

**Comment by Opti98** [comment]
We generate out certificates from another authority too, and when we replace the certificates Traefik doesn't see the change. I have discovered if I touch the configuration file it does reload the new certificates, ie  "touch certificates.toml"

**Comment by Sarke** [comment]
> I have discovered if I touch the configuration file it does reload the new certificates, ie "touch certificates.toml"

Really? That did not work for me last time I tried.  I think Traefik caches the parsed config files and if it's still the same it doesn't re-apply them.

**Comment by AndrewSav** [comment]
Of course traefik reloads the config file if it monitors it. This issue is for the case when inotify is not available, e.g. remote mount or when you mount a single file with docker. The current functionality covers vast majority of use cases, but not all.

**Comment by Ajedi32** [comment]
After encountering this problem myself I've done a bit of experimentation, and there seem to be several issues at play here:

First, it appears **Traefik doesn't monitor the certificates files for changes at all**. I might be wrong about this (I haven't dug around in the source code or anything) but based on my experiments (running Traefik 2.3.2 via the traefik:v2.3 image on Windows Subsystem for Linux) Traefik doesn't watch for changes to the certificate or key files. It _does_, however, watch for changes to the dynamic config file, in cases where you're using the file provider. If you run `touch $file` against your dynamic config file, Traefik will then check whether you made changes to the certificate or key files referenced in that config.

Secondly, the above trick where you `touch` the dynamic config file to reload certificates only works if the certificate in question is referenced under the `tls.certificates` key. It does _not_ work for certificates that are _only_ listed in the default certificate store. E.g. If you have this config:

```
tls:
  stores:
    default:
      defaultCertificate:
        certFile: /etc/traefik/certificates/my-site.cer
        keyFile: /etc/traefik/certificates/my-site.key
```

Then TLS will work as expected, but Traefik will _not_ reload changes to the key or certificate files _even_ if you `touch` the dynamic config file. No idea why. In order for that trick to work, you'd need a config more like this:

```
tls:
  certificates:
    - certFile: /etc/traefik/certificates/my-site.cer
      keyFile: /etc/traefik/certificates/my-site.key

  stores:
    default:
      defaultCertificate:
        certFile: /etc/traefik/certificates/my-site.cer
        keyFile: /etc/traefik/certificates/my-site.key
```

Third, note that even if Traefik _does_ reload the certificates, those changes may not immediately affect the web browser you're testing with if that browser still has an open connection with the Traefik server. In order for the changes to take effect you may need to restart the browser to force it to close any existing connections. (In Chrome, pressing the "Close idle sockets" button in chrome://net-internals/#sockets will also work.)

**Comment by Gallardo994** [comment]
I am currently deploying Traefik into kubernetes with cert-manager providing wildcard certificates. Wildcard cert is used as default in Traefik via secret mounted together with configmap. According to this issue, should I be concerned with "restarting traefik" each time cert-manager renews the certs?

**Comment by redondos** [comment]
> I am currently deploying Traefik into kubernetes with cert-manager providing wildcard certificates. Wildcard cert is used as default in Traefik via secret mounted together with configmap. According to this issue, should I be concerned with "restarting traefik" each time cert-manager renews the certs?

We run Traefik 1.7 and I don't know the 2.x behavior. Instead of restarting traefik pods to reload certificate secrets we rely on [Global Default Backend](https://doc.traefik.io/traefik/v1.7/configuration/backends/kubernetes/#global-default-backend-ingresses) ingress rules to overcome this limitation.

We create a `*` ingress rule in a namespace watched by traefik that includes a `tls` section with all the wildcards we need. cert-manager will manage those certificates and secrets, and traefik will use them for TLS termination.

**Comment by zedtux** [comment]
No matter the cases where Traefik woud or wouldn't reload the certificates, the initial request is about a way to force Traefik to reload the certs.

Is there any plans to implement it since the year and some months since this issue has been opened?

**Comment by bkraul** [comment]
Wondering the same thing. We are getting ready to deploy traefik in production with a large number of domains/certs...services we can't really stop, so a reload for SSL certs is a must...apache and nginx provide this functionality.

**Comment by tsarna** [comment]
More specifically, I would like that way to be by sending a signal, such as SIGHUP

**Comment by bkraul** [comment]
A reload option would be awesome tho. SIGHUP would mean the service is being restarted. The point is to have no interruption. Take for instancd `nginx -s reload` or `httpd reload` neither of them disrupt service to the running web application. But they do update most configuration settings on the fly...

**Comment by tsarna** [comment]
@bkraul SIGHUP does not mean the service is being restarted. It's very common for daemons to catch SIGHUP as an indication that configuration should be reloaded. Apache, Nginx, sendmail, BIND, dhcpd, traditional implementations of inetd and cron, etc. all implement configuration reload on SIGHUP, and thus it's well supported by the kind of tools often used for updating configuration files. Although most tools support configuring a different signal today, in my experience SIGHUP is the most commonly implemented.

**Comment by MartinMuzatko** [comment]
@bkraul  @tsarna how do you do sighup via docker? wouldn't an API be better?

**Comment by florentchauveau** [comment]
Hello all,

@Ajedi32 posted the solution above. It works.

TLDR: **traefik does not monitoring the certificate files, it monitors the dynamic config file**

Steps:
1. Update your cert file
2. Touch dynamic.yml
3. Et voilà, traefik has reloaded the cert file

There might be a gotcha with the default certificate store. You have to list your certificates twice. My dynamic.yml file looks like this:
```
tls:
  certificates:
    - certFile: /etc/ssl/any.site.com.crt
      keyFile: /etc/ssl/any.site.com.key
  stores:
    default:
      defaultCertificate:
        certFile: /etc/ssl/any.site.com.crt
        keyFile: /etc/ssl/any.site.com.key
```

You can test with `curl -v`, which avoids browser cache.

**Comment by tsarna** [comment]
> @bkraul how do you do sighup via docker?

In Nomad, I would just do something like:

```
      template {
        data = <<-__EOF__
        {{ with secret "pki/issue/..." "common_name=..." ... }}
        {{ .Data.private_key }}
        {{ end -}}
        __EOF__

        destination   = "${NOMAD_SECRETS_DIR}/key.pem"
        change_mode   = "signal"
        change_signal = "SIGHUP"
      }
```

I'm assuming there's a way to do something similar with other container orchestrators.

> @bkraul wouldn't an API be better?

I'd have to write some integration code to use that API, maintain that code, source control it somewhere, deploy the code, dedicate resources to running the code, think about the security of that code and about securing the API...

**Comment by mars64** [comment]
In a kubernetes context, the [stakater/reloader](https://github.com/stakater/Reloader) project might be interesting here. Just hook it up to the secret object that represents your certificate and configure traefik for multi-replica.

I'd still prefer a reload endpoint in traefik over rotating the pod entirely, but this seems a reasonable interim solution.

**Comment by zedtux** [comment]
Nice suggestion @mars64 👍

**Comment by cinderblockgames** [comment]
> Hello all,
> 
> @Ajedi32 posted the solution above. It works.
> 
> TLDR: **traefik does not monitoring the certificate files, it monitors the dynamic config file**
> 
> Steps:
> 
> 1. Update your cert file
> 2. Touch dynamic.yml
> 3. Et voilà, traefik has reloaded the cert file
> 
> There might be a gotcha with the default certificate store. You have to list your certificates twice. My dynamic.yml file looks like this:
> 
> ```
> tls:
>   certificates:
>     - certFile: /etc/ssl/any.site.com.crt
>       keyFile: /etc/ssl/any.site.com.key
>   stores:
>     default:
>       defaultCertificate:
>         certFile: /etc/ssl/any.site.com.crt
>         keyFile: /etc/ssl/any.site.com.key
> ```
> 
> You can test with `curl -v`, which avoids browser cache.

This doesn't work for me.  I have the keys loaded via dynamic file:
```
[tls]
  [[tls.certificates]]
    certFile = "/config/certs/wildcard.homelab.express.cer"
    keyFile = "/config/certs/wildcard.homelab.express.key"
    stores = ["default"]

  [tls.stores]
    [tls.stores.default]
      [tls.stores.default.defaultCertificate]
        certFile = "/config/certs/wildcard.homelab.express.cer"
        keyFile = "/config/certs/wildcard.homelab.express.key"
```
I replace the files, touch certs.toml, and nothing.  I've tried pulling up a different browser entirely to make sure nothing is cached, but I'm not getting the updated cert anywhere.  I even tried doing a docker container restart on traefik; still no joy.

**Comment by bkraul** [comment]
Yup this does not work for me with touching 2.4 `dynamic.yml` either.

**Comment by cinderblockgames** [comment]
> > Hello all,
> > @Ajedi32 posted the solution above. It works.
> > TLDR: **traefik does not monitoring the certificate files, it monitors the dynamic config file**
> > Steps:
> > 
> > 1. Update your cert file
> > 2. Touch dynamic.yml
> > 3. Et voilà, traefik has reloaded the cert file
> > 
> > There might be a gotcha with the default certificate store. You have to list your certificates twice. My dynamic.yml file looks like this:
> > ```
> > tls:
> >   certificates:
> >     - certFile: /etc/ssl/any.site.com.crt
> >       keyFile: /etc/ssl/any.site.com.key
> >   stores:
> >     default:
> >       defaultCertificate:
> >         certFile: /etc/ssl/any.site.com.crt
> >         keyFile: /etc/ssl/any.site.com.key
> > ```
> > 
> > 
> > You can test with `curl -v`, which avoids browser cache.
> 
> This doesn't work for me. I have the keys loaded via dynamic file:
> 
> ```
> [tls]
>   [[tls.certificates]]
>     certFile = "/config/certs/wildcard.homelab.express.cer"
>     keyFile = "/config/certs/wildcard.homelab.express.key"
>     stores = ["default"]
> 
>   [tls.stores]
>     [tls.stores.default]
>       [tls.stores.default.defaultCertificate]
>         certFile = "/config/certs/wildcard.homelab.express.cer"
>         keyFile = "/config/certs/wildcard.homelab.express.key"
> ```
> 
> I replace the files, touch certs.toml, and nothing. I've tried pulling up a different browser entirely to make sure nothing is cached, but I'm not getting the updated cert anywhere. I even tried doing a docker container restart on traefik; still no joy.

Nevermind -- I still had my traefik volume mounted against the wrong location.  This does work for me.

**Comment by gseidel** [comment]
I also would like to see this feature more soon as late. I'm running into the problem, that I adding host/routers to traefik before I change the DNS record for migration and testing reasons. So of course the acme challenge will fail. After I change my DNS Record to the traefik server it will not do the challenge again. The only solution to trigger that, is to remove and add my traefik docker, causing downtime to all other hosts.

**Comment by microadam** [comment]
I also can't get @Ajedi32 solution to work. Slightly different scenario in that I am trying to use client certificates for the backends, which change reguarly.

```
[http.serversTransports.nomad-client]
  serverName = "client.cluster.nomad"
  rootCAs = [ "/etc/nomad.d/nomad-ca.pem" ]
  [[http.serversTransports.nomad-client.certificates]]
    certFile = "/etc/nomad.d/nomad.crt"
    keyFile = "/etc/nomad.d/nomad.key"
```

When the certificate changes (stored in vault, and templated out by vault-agent), a `touch` of the dynamic config file does indeed reload the config (can see from Traefik logs) however the SSL cert isn't reloaded until Traefik is restarted.

I have tried adding the `certFile` and `keyFile` under `[[tls.certificates]]` but that doesn't seem to have any effect.

Anyone have any suggestions? workarounds? Or is there any feature upcoming in the roadmap at all? (API endpoint or SIGHUP etc to reload certs). Restarting traefik is far from ideal

**Comment by microadam** [comment]
Hi @jbdoumenjou I hope you don't mind me pinging you on this. The issue hasn't had any input from any maintainers for some time, so it would be great to understand if this is a bug, or if what has been suggested previously SHOULD work, or if there is scope for some new feature additions.

**Comment by jbdoumenjou** [comment]
Hi @microadam,

The configuration itself does not change when the certificates are updated, so the configuration update must be forced.
In that case, the touch of the dynamic conf is the best way to force the configuration update. 

Currently, the [TLS certificates defined by user](https://doc.traefik.io/traefik/https/tls/#user-defined) will be loaded at each configuration update as we read the certificates to check if they have changed.

Thanks to your comment, we saw that this does not work for the other certificates' configurations ([Certificates Stores](https://doc.traefik.io/traefik/v2.4/https/tls/#certificates-stores), [TLS Options](https://doc.traefik.io/traefik/v2.4/https/tls/#tls-options) and [serversTransport](https://doc.traefik.io/traefik/v2.4/routing/services/#serverstransport_1)).
So, we open #8243. Feel free to interact/test it. :wink:

**Comment by microadam** [comment]
@jbdoumenjou amazing! Thank you very much for getting back to me and for working out and fixing the problem! I will figure out how to compile from source and give that branch a whirl :)

**Comment by houzhx759** [comment]
Hi, have you solved this problem yet?I have now also encountered a replacement traefik certificate that does not take effect,My service certificate is about to expire,I did this by manually deleting Secret,Then I recreate secret TLS but it doesn't work,Even weirder is that sometimes it works,But it will soon be back to the old certificate.This problem has bothered me for several days and I hope to reply. thank you

**Comment by zerodayyy** [comment]
> In a kubernetes context, the [stakater/reloader](https://github.com/stakater/Reloader) project might be interesting here. Just hook it up to the secret object that represents your certificate and configure traefik for multi-replica.
> 
> I'd still prefer a reload endpoint in traefik over rotating the pod entirely, but this seems a reasonable interim solution.

Implemented this approach in a production environment, works flawlessly. Seems like a decent workaround until this issue is solved.

**Comment by xzycn** [comment]
Hi, is there any update for this?
I‘m using cert-manager to manage certificates, and file provider for traefik(2.6.1)，certificate has been updated，but  the previous one still in use. I don't want traefik to restart :)

**Comment by dginhoux** [comment]
Hi, 

I've the same problem here when everything is stored on NFS share provided by netapp.
It could be very great to have an api endpoint that force all config refresh.

Have a good day.

**Comment by farzadso** [comment]
> Same here with traefik 1.x helm chart. Using existing secret for bought tls certs in kubernetes and telling traefik via ingress rule what the secretname is. Updating the secret has no effect at all. The old cert is still used...
> 
> Edit:
> 
> Seems our problem was that we had several secrets wiht the same name "my-tls-wildcard-secret" in different namespaces. We had to update all of them before traefik used it.

This was my issue too. Took me a couple of days to find your useful reply. Much appreciated.

It's strange how the Ingress would pick up certificates that weren't bound to it.

**Comment by avluis** [comment]
👀

**Comment by bkraul** [comment]
There is also the issue of docker swarm setup. Let's say you have 3 manager nodes, you can touch the cert config yml on one of them, and it will work great, but the other two will know nothing about the updated cert...hence, issues. Same with acme. Only thing resolving it is a restart of the service stack.

**Comment by jralph** [comment]
> 

Also just experienced this issue. Its not based on the secret name, but the primary domain name I believe. We had 3 secrets with different names, 2 had expired and weren't referenced anywhere anymore.

Some how, even after a reload of traefik, it was still serving these old certificates instead of the new one (we have 3 nodes running, 1 was serving the old cert, 2 the new).

It really does feel odd that traefik serves a cert that the ingressroute doesn't ever reference or tell it to, even after a full restart.

**Comment by mpl** [comment]
Hello @farzadso, @jralph,

just so you know, we've hidden your comments, as they were off-topic, to keep the issue on track and to avoid misleading other readers of the issue.

**Comment by Amunak** [comment]
@Ajedi32 is spot on, also about the cert configured in default certs only not being enough.

What is absolutely baffling to me is that Traefik devs are so ignorant towards implementing a simple signal for reloading the configuration as pretty much every other daemon does. Sure, touching a dynamic config file works, but it's a horrible workaround - next time they'll implement configuration caching so only changes propagate and it'll stop working or something.

I'm on Traefik v3 now and #8243 doesn't seem to solve it for me (I have the certs mounted in a directory and not separately), a reload or restart is still needed. But my testing was limited as it's hard to test properly without accidentally reloading the configuration.


> how do you do sighup via docker? wouldn't an API be better?

@MartinMuzatko you can do something like `docker kill -s SIGUSR1 traefik`. But it's not mutually exclusive either; I think it'd be nice if we got both an API and a signal trap. The API would almost certainly be more work though, especially since it'd be the first write API which would require some extra considerations to start with.

**Comment by Sarke** [comment]
> There is also the issue of docker swarm setup. Let's say you have 3 manager nodes, you can touch the cert config yml on one of them, and it will work great, but the other two will know nothing about the updated cert...hence, issues. Same with acme. Only thing resolving it is a restart of the service stack.

@bkraul I ended up doing a touch on the config files, then rsync to the other nodes.  Then they all pick up the changes.

**Comment by bkraul** [comment]
Sounds legit...though a bit hacky. At that point I might just set up an Ansible playbook to do just what you described.

**Comment by nmengin** [comment]
Hello,

We have recently merged a [PR](https://github.com/traefik/traefik/pull/9993) that allows reloading the configuration when Traefik receives a SIGHUP signal.
Such a feature allows Traefik users to reload manually (executing the command `kill -1 <TraefikProcessId>`) when the TLS certificates are updated.
This feature will be available in the next version v3.0.

Is it enough for your use cases?

**Comment by bkraul** [comment]
Na man...endpoint call would have been nice.

**Comment by AndrewSav** [comment]
Not to mention that Windows does not have SIGHUP

**Comment by ybizeul** [comment]
Just trying to sum this up :
- SIGHUP not available until 3.0
- Dynamic config requires an actual change of the conf file, changing the certificates on disk won't trigger a refresh, neither would `touch` the config files :
```
time="2024-03-15T17:29:33Z" level=debug msg="Skipping unchanged configuration." providerName=file
```

I'm sure that's useful for some use case, but definitely leaves a lot of people aside that were hoping for a simple monitoring of the cert files

**Comment by rtribotte** [comment]
Hello,

A configurable HTTP endpoint to trigger configuration reloads is a good enhancement.
We think it should be an internal service, like `api@internal`.
It would not be possible to expose it via an option (e.g.: `api.insecure`) but only with a router, and its enablement would be controlled by a static configuration option.

We would love some help from the community on this, if any community member would like to contribute to this, let us know, and we will work with you to make sure you have all the information needed before starting, and ensure that we are aligned and can move quickly with the review and merge process.

**Comment by D0wn3r** [comment]
Hi !
I'm looking for the same feature that will reload well my certificates.
my dynamic configuration:
```
[http]
  [http.middlewares]
    [http.middlewares.vpn-ipwhitelist.ipWhiteList]
      sourceRange = ["10.0.0.0/8", "172.80.0.0/16", "172.90.0.0/16", "172.18.0.0/16"]
[tls.options]
  [tls.options.default]
    minVersion = "VersionTLS12"
    cipherSuites = [
      "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
      "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
      "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305",
      "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305"
    ]
[[tls.certificates]]
  certFile = "/etc/le/fullchain.pem"
  keyFile  = "/etc/le/privkey.pem"


[[tls.certificates]]
  certFile = "/etc/le/fullchain_pca12.pem"
  keyFile  = "/etc/le/privkey_pca12.pem"
```

When I change it or just touch the file, I have this log
```
level=debug msg="Configuration received: {\"http\":{\"middlewares\":{\"vpn-ipwhitelist\":{\"ipWhiteList\":{\"sourceRange\":[\"10.0.0.0/8\",\"172.80.0.0/16\",\"172.90.0.0/16\",\"172.18.0.0/16\"]}}}},\"tcp\":{},\"udp\":{},\"tls\":{\"options\":{\"default\":{\"minVersion\":\"VersionTLS12\",\"cipherSuites\":[\"TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256\",\"TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256\",\"TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384\",\"TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384\",\"TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305\",\"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305\"],\"clientAuth\":{},\"alpnProtocols\":[\"h2\",\"http/1.1\",\"acme-tls/1\"]}}}}" providerName=file
time="2024-04-04T17:26:40+02:00" level=debug msg="Skipping unchanged configuration." providerName=file
```
Is that normal than TLS certificate are not mentionned in the "Configuration received" ?

**Comment by D0wn3r** [comment]
> Yes, an endpoint can be great.
> 
> More simple... a flag ! Watch for a file "/certs/.need_to_update", if present, start update process and remove the flag.

Not great with multiple Traefik instances using the same folder of certificates. Only 1 will be able to update

**Comment by cdwiegand** [comment]
I looked around the code, and I'm not super-golang-capable, but I think with the way Providers are separate from DynamicConfig that it'd be hard to implement a way to ask the FileProvider to reload itself from the api package, as the api doesn't have access to the AggregateProvider (or its internal FileProvider instance), and FileProvider doesn't expose a way to reload its configuration externally - just fsnotify and at load time. 

I did consider adding a time.Ticker routine and reevaluate every watched file every x seconds, but didn't really like the resulting code. 

I also considered seeing if the RestProvider could "call in" with its own channel to AggregateProvider to "ask FileProvider, if valid, to reload itself" as AggregateProvider does have a reference to it (if using that provider), but I also felt that was very hacky and definitely didn't fit the way traefik is architected.

I ended up building a docker image with supervisord, with XML-RPC enabled, so I can call the endpoint there to SIGHUP the traefik process. It isn't ideal, but it's a valid workaround until someone can implement an API endpoint directly in traefik.
